### PR TITLE
✨ insert tree

### PR DESCRIPTION
**Pull request by BoonChi**
image

🐛 not able to create consecutive subtree with current code

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,85 @@
+interface ITree {
```
Comment (Dawidpol) [inline index.ts:1]: Let's make the tree generic 
```
interface ITree<TValue = any>
```

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  left: ITree;
+  right: ITree;
+  level: number;
+  constructor() {
+    this.value = undefined;
+    this.left = {
+      value: undefined,
+      left: undefined,
+      right: undefined
+    };
```
Comment (Dawidpol) [inline index.ts:23]: Why we initiate with left and right value set?

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  right?: ITree;
+}
+
+enum TreeDirection {
+  left = 'left',
+  right = 'right'
+}
+
+class Tree implements ITree {
+  value: number | undefined;
```
Comment (Dawidpol) [inline index.ts:13]:
```suggestion
  value?: number;
```

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      value: undefined,
+      left: undefined,
+      right: undefined
+    };
+    this.right = {
+      value: undefined,
+      left: undefined,
+      right: undefined
+    };
+    this.level = 0;
```
Comment (Dawidpol) [inline index.ts:29]: We shouldn't keep the level as this would gratly reduce the insert speed

File: index.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+enum TreeDirection {
+  left = 'left',
+  right = 'right'
+}
+
+class Tree implements ITree {
+  value: number | undefined;
+  left: ITree;
+  right: ITree;
+  level: number;
```
Comment (Dawidpol) [inline index.ts:16]: I think instead of so many classes we can just implemnt the tree on simple js object
```suggestion
root: ITree
```

Then to make it more readalbe let's change the name `ITree` to `ITreeNode`

**Review by Dawidpol — CHANGES_REQUESTED** [review]

File: index.ts (hunk 5 above)
Comment (BoonChi) [reply index.ts:16]: you suggested not to use class at all?

File: index.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        value,
+      }
+    }
+  }
+
+  private _insert(node: ITreeNode<TValue>, value: TValue) {
+    let direction = 'left'
+    if (value >= node.value) {
+      direction = 'right'
+    }
```
Comment (Dawidpol) [inline index.ts:25]: You can make it look better

```suggestion
const direction = value >= node.value ? 'right' : 'left'
```

File: index.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+}
+// test begins
+const a = new Tree<number>();
+a.insert(10)
+a.insert(8)
+a.insert(20)
+a.insert(31)
+a.insert(9)
+console.log(JSON.stringify(a))
+// test ends
```
Comment (Dawidpol) [inline index.ts:44]: We should add the unit test instead

File: index.ts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+      direction = 'right'
+    }
+    if (node[direction] !== undefined) {
+      this._insert(node[direction], value)
+    } else {
+      node[direction] = {
+        value
+      }
+    }
+  }
```
Comment (Dawidpol) [inline index.ts:33]: You can simplify it a bit
```
if(node[direction]) this._insert(node[direction], value)
else node[direction] = {value}
```

**Review by Dawidpol — CHANGES_REQUESTED** [review]

**Comment by Dawidpol** [comment]
@BoonChi Almost there just add unit tests and apply comments above. the next step is simple tree remove

**Comment by BoonChi** [comment]
> @BoonChi Almost there just add unit tests and apply comments above. the next step is simple tree remove

unit test with coverage 100% is added and all comments are resolved.

File: index.test.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+  test("should insert under right sub tree(leaf node)", () => {
+    const treeNode = new Tree();
+    treeNode.insert(3)
+    treeNode.insert(1)
+    treeNode.insert(5)
+    treeNode.insert(10)
+    expect(treeNode.root).toStrictEqual({ value: 3, left: { value: 1 }, right: { value: 5, right: { value: 10 } } });
+  });
+});
```
Comment (Dawidpol) [inline index.test.ts:33]: Need to add eslint to make sure we have ne line at the end of the file, but can do this in separate PR

**Review by Dawidpol — APPROVED** [review]
